Mobile drafts

Rajat-Dabade opened the pull request:
#### Summary
Separate tab for Drafts in mobile. 

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-39356

#### Checklist
<!--
Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.
-->
- [ ] Added or updated unit tests (required for all new features)
- [x] Has UI changes
- [ ] Includes text changes and localization file updates
- [ ] Have tested against the 5 core themes to ensure consistency between them.
- [ ] Have run E2E tests by adding label `E2E iOS tests for PR`.

#### Device Information
This PR was tested on: iOS version 17.5 (Both on iPhone and ipad)

#### Screenshots
<!--
If the PR includes UI changes, include screenshots/GIFs/Videos (for both iOS and Android if possible).
-->

#### Release Note
```release-note
Added a feature for having a separate tab for Drafts in mobile. 
```

Note: 
Will cover all the test case for this PR in separate PR.

Rajat-Dabade commented:
/update-branch

larkox reviewed (changes requested):
Great work! Some comments.

Rajat-Dabade commented on app/components/draft/draft.tsx:
```diff
@@ -0,0 +1,111 @@
+    const isTablet = useIsTablet();
+    const showPostPriority = Boolean(isPostPriorityEnabled && draft.metadata?.priority && draft.metadata?.priority?.priority);
+
+    const onLongPress = () => {
```
Done.

Rajat-Dabade commented on app/components/draft/draft_post/draft_files/index.ts:
```diff
@@ -0,0 +1,23 @@
+
+    return {
+        canDownloadFiles: observeCanDownloadFiles(database),
+        publicLinkEnabled,
```
Done.

Rajat-Dabade commented on app/components/draft/draft_post/draft_message/index.tsx:
```diff
@@ -0,0 +1,105 @@
+    return (
+        <>
+            <Animated.View style={animatedStyle}>
+                <ScrollView
+                    keyboardShouldPersistTaps={'always'}
+                    scrollEnabled={false}
+                    showsVerticalScrollIndicator={false}
+                    showsHorizontalScrollIndicator={false}
+                >
```
When the content is wrapped in a ScrollView, React Native can measure the full height of the content accurately, even if the content is overflowing. This is because ScrollView enables scrolling, so it doesn't limit the height calculation to what's immediately visible. Without ScrollView, React Native would only measure the visible part, often resulting in a shorter, incorrect height measurement. This may result in not showing the ShowMore button.

Rajat-Dabade commented on app/components/draft/index.tsx:
```diff
@@ -0,0 +1,94 @@
+    };
+});
+
+export default React.memo(
+    withDatabase(
+        withChannel(
+            withCurrentUser(
+                withChannelMembers(
```
This was just to split the code and make it more readable. Can we revertable.

Rajat-Dabade commented on app/components/drafts_buttton/drafts_button.tsx:
```diff
@@ -0,0 +1,136 @@
+
+import CompassIcon from '../compass_icon';
+
+// See LICENSE.txt for license information.
```
Done.

Rajat-Dabade commented on app/components/drafts_buttton/drafts_button.tsx:
```diff
@@ -0,0 +1,136 @@
+        </TouchableOpacity>
+    );
+
+    // const renderItem = ({item, index}: {item: string; index: number}) => (
+    //     <View style={styles.itemContainer}>
+    //         <Text style={styles.messageText}>{'Draft'} {index + 1} {':'} {item}</Text>
+    //         <Text style={styles.fileText}>{'Attached files:'} {files[index]?.length || 0}</Text>
+    //     </View>
+    // );
+
+    // return (
+    //     <FlatList
+    //         data={messages}
+    //         keyExtractor={(item, index) => `draft-${index}`}
+    //         renderItem={renderItem}
+    //         ListEmptyComponent={<Text style={styles.emptyText}>{'No drafts available.'}</Text>}
+    //     />
+    // );
```
Done.

Rajat-Dabade commented on app/queries/servers/drafts.ts:
```diff
@@ -30,3 +29,27 @@ export const queryDraft = (database: Database, channelId: string, rootId = '') =
+            Q.and(
+                Q.or(
+                    Q.where('team_id', teamId), // Channels associated with the given team
+                    Q.where('type', 'D'), // Channels of type 'D'
+                    Q.where('type', 'G'), // Channels of type 'D'
```
Yup, your suggestion makes more sense. Updated it.

Rajat-Dabade commented on assets/base/i18n/en.json:
```diff
@@ -173,6 +173,7 @@
   "channel_info.favorited": "Favorited",
   "channel_info.header": "Header:",
   "channel_info.ignore_mentions": "Ignore @channel, @here, @all",
+  "channel_info.in": "In:",
```
Sure.

Rajat-Dabade commented on assets/base/i18n/en.json:
```diff
@@ -305,6 +308,16 @@
+  "draft.options.edit.title": "Edit draft",
+  "draft.options.send.title": "Send draft",
+  "draft.options.title": "Draft Options",
+  "drafts": "Drafts",
+  "Drafts": "Drafts",
```
Removed `Drafts` one.

Rajat-Dabade commented on app/components/post_draft/post_input/post_input.tsx:
```diff
@@ -144,10 +144,15 @@ export default function PostInput({
 
-    const onBlur = useCallback(() => {
+    const onBlur = useCallback(async () => {
         updateDraftMessage(serverUrl, channelId, rootId, value);
```
Yup, we should wait now in order to create draft first before parsing the image.

Rajat-Dabade commented on app/actions/local/draft.ts:
```diff
@@ -197,3 +211,94 @@ export async function updateDraftPriority(serverUrl: string, channelId: string,
+    try {
+        const {database, operator} = DatabaseManager.getServerDatabaseAndOperator(serverUrl);
+        const draft = await getDraft(database, channelId, rootId);
+        if (!draft) {
+            const newDraft: Draft = {
+                channel_id: channelId,
+                root_id: rootId,
+                metadata: {
+                    images: imageMetadata,
+                },
+                update_at: Date.now(),
+            };
+
+            return operator.handleDraft({drafts: [newDraft], prepareRecordsOnly});
+        }
```
Yup, it's a nice catch, whenever we are adding the image metadata, the draft record is already created and no need to create a new draft for it (as we are not waiting for `updateDraftMessage`: https://github.com/mattermost/mattermost-mobile/pull/8280#discussion_r1830813027

yasserfaraazkhan reviewed (changes requested):
Thank you @Rajat-Dabade 

Most flows around 
- Creating Draft in DM, GM, Channel, Threads, Private Chanel, under deleted thread  ✅ 
- Sending Threads, Editing threads work as expected. ✅ 

There couple of flow I wanted to bring to your notice
1. On the Draft Screen, if we hit on Android's back button _(not the back arrow of MM app)_ the app closes instead of going back to channels list page

2. Once all drafts are sent out from the draft page, we have an empty page looking like below. would it be nice to add a 'No drafts at the moment' icon, like in web app. @asaadmahmood thoughts?

![Screenshot_20241118-143836](https://github.com/user-attachments/assets/e2252059-6392-4e84-aefd-397376581c99)

Rajat-Dabade commented:
@yasserfaraazkhan 
- On the Draft Screen, if we hit on Android's back button (not the back arrow of MM app) the app closes instead of going back to channels list page ✅ 
- Once all drafts are sent out from the draft page, we have an empty page looking like below. would it be nice to add a 'No drafts at the moment' icon, like in web app. @asaadmahmood thoughts? ✅ 

Can you please review again? Thanks.

yasserfaraazkhan reviewed (changes requested):
Thank you @Rajat-Dabade 
just 1 last change, in below flow 

**Scenarios 1:**
Steps:
- From Draft page, Edit a draft. Verify you are taken to the Channel/DM/GM
- Edit the Draft in the Channel
- Click on the `<` back Icon on the channel

This is taking user to `Draft page`. should it take user to Channels List page?

**Scenarios 2:**

With above steps, if we edit draft message _(without sending it)_ and Click back button, the draft message is not updated.

https://github.com/user-attachments/assets/78a21082-3042-489a-b91d-b72df9c1db7b

**Scenario 3:**
- The User group Name is shown like below

<img width="377" alt="Screenshot 2024-11-19 at 12 29 22 AM" src="https://github.com/user-attachments/assets/efdf05a3-0476-47dd-a4c4-e82f7441a993">

on web app it show as below

<img width="656" alt="image" src="https://github.com/user-attachments/assets/64137399-c1d6-4979-b8b3-292692b15f5f">

asaadmahmood reviewed (changes requested):
Overall looks good to me, minor changes though.
1. ![IMG_0360](https://github.com/user-attachments/assets/337decea-ac44-4326-bd2d-c726363ae04b)
2. ![IMG_0358](https://github.com/user-attachments/assets/7507d378-c6a3-4105-9f0f-8c4bae6d39f9)
3. https://www.loom.com/share/37c3dde4ec484cd2a61d5afe00bb1657?from_recorder=1&focus_title=1

asaadmahmood commented:
@Rajat-Dabade Also one minor thing, clicking on the message in the drafts view should take you to the channel where that message is in drafts.

asaadmahmood commented:
Also @Rajat-Dabade the first time the user goes to the draft screen (with a single draft), we show this to him.
Here's the Figma link:
https://www.figma.com/design/VbEx15Tpvi9mXhs9IEjJtq/MM-45558-Drafts-%26-Schedule-Messages?node-id=5410-97500&node-type=frame&t=mL9UIcpxk6XmrU3V-11

![image](https://github.com/user-attachments/assets/14fd77f2-afb8-43a5-9eaf-8bf106af55f1)

larkox commented on app/screens/draft_options/send_draft/index.tsx:
```diff
@@ -0,0 +1,91 @@
+    channelId: string;
+} & WithDatabaseArgs;
+
+const enhanced = withObservables([], ({database, rootId, channelId}: Props) => {
```
0/5 I wonder if this can be shared with the post input component.

That being said... probably they are different enough to better keep them separate.

Rajat-Dabade replied:
Yup, it is better to keep them separate because of their tightly bound nature.

enahum replied:
What makes it tightly bound?

larkox reviewed (changes requested):
Great work! Many comments, but mostly nitpicks.

Rajat-Dabade commented on app/components/channel_info/channel_info.tsx:
```diff
@@ -0,0 +1,170 @@
+            color: changeOpacity(theme.centerChannelColor, 0.64),
+            ...typography('Body', 75, 'SemiBold'),
+        },
+        timeInfo: {
+            fontSize: 12,
+            fontWeight: '400',
+            color: changeOpacity(theme.centerChannelColor, 0.64),
+            lineHeight: 16,
```
This is not used anywhere, I forgot to delete it. Nice catch. Will remove it.

Rajat-Dabade commented on app/components/channel_info/avatar/index.tsx:
```diff
@@ -0,0 +1,71 @@
+    const serverUrl = useServerUrl();
+
+    let uri = '';
+    if (!uri && author) {
```
Sure, removed it.

Rajat-Dabade commented on app/queries/servers/drafts.ts:
```diff
@@ -30,3 +29,27 @@ export const queryDraft = (database: Database, channelId: string, rootId = '') =
+            Q.and(
+                Q.or(
+                    Q.where('team_id', teamId), // Channels associated with the given team
+                    Q.where('type', 'D'), // Direct Message
+                    Q.where('type', 'G'), // Group Message
```
The `type` approach looks cleaner and more informative, IMHO. Also, hint to the future developer that this code is for Direct/group related only.

Rajat-Dabade commented on app/queries/servers/drafts.ts:
```diff
@@ -30,3 +29,27 @@ export const queryDraft = (database: Database, channelId: string, rootId = '') =
+};
+
+export const observeAllDrafts = (database: Database, teamId: string) => {
+    return queryAllDrafts(database, teamId).observeWithColumns(['messages', 'files', 'metadata']);
```
Yup, that's correct, now that we are updating `update_at` every time we are updating the draft, better observe that column.

Updated it.

Rajat-Dabade commented on app/screens/draft_options/delete_draft/index.tsx:
```diff
@@ -0,0 +1,72 @@
+    const style = getStyleSheet(theme);
+    const serverUrl = useServerUrl();
+
+    const draftDeleteHandler = async () => {
```
We do have delete confirm for the swipe delete draft, but let me confirm this for this handler.

enahum replied:
Was this confirmed?

Rajat-Dabade replied:
Yup, this was confirmed. We have to add the confirmation modal for deleting the drafts.  Added it.

Rajat-Dabade commented on app/screens/draft_options/index.tsx:
```diff
@@ -0,0 +1,91 @@
+            componentId={Screens.DRAFT_OPTIONS}
+            renderContent={renderContent}
+            closeButtonId={DRAFT_OPTIONS_BUTTON}
+            snapPoints={[200, 250]}
```
It just has three options as of now, so I guess it's okay to keep it as magic numbers.

enahum replied:
I'm with Dan here, why magic numbers and in any case is missing one to when we close it

Rajat-Dabade commented on assets/base/images/Draft_Message.png:
I am open if we want to go with SVGs but in many places in the application we have used PNGs as well. 

![Screenshot 2024-12-05 at 6 01 08 PM](https://github.com/user-attachments/assets/b7dd005c-1ca3-4125-8bd4-f554f9787169)

Rajat-Dabade commented on app/screens/global_drafts/components/global_drafts_list/global_drafts_list.tsx:
```diff
@@ -0,0 +1,160 @@
+                setTooltipVisible(true);
+            }
+        });
+    }, [tutorialWatched]);
```
Yup, that's correct. ermoved it.

Rajat-Dabade commented on app/screens/global_drafts/components/global_drafts_list/global_drafts_list.tsx:
```diff
@@ -0,0 +1,160 @@
+    }), [isTablet]);
+
+    useEffect(() => {
+        InteractionManager.runAfterInteractions(() => {
+            if (!tutorialWatched) {
+                setTooltipVisible(true);
+            }
+        });
```
Done.

Rajat-Dabade commented on app/screens/global_drafts/components/global_drafts_list/global_drafts_list.tsx:
```diff
@@ -0,0 +1,160 @@
+    }, []);
+
+    const renderItem = useCallback(({item}: ListRenderItemInfo<DraftModel>) => {
+        if (item.id === firstDraftId && tutorialWatched) {
```
Yup, typo for testing. Reverted it back.

Rajat-Dabade commented on app/screens/global_drafts/index.tsx:
```diff
@@ -0,0 +1,97 @@
+        return {flex: 1, marginTop};
+    }, [defaultHeight]);
+
+    const onBackPress = useCallback(() => {
```
Yup, we are using `useAndriodHardwareBackHandler` but here: `app/screens/global_drafts/components/global_drafts_list/global_drafts_list.tsx`.

Rajat-Dabade commented on app/components/drafts_buttton/index.tsx:
```diff
@@ -0,0 +1,31 @@
+}));
+
+const enhanced = withObservables(['teamId'], ({database, teamId}: Props) => {
+    const draftsCount = observeDraftCount(database, teamId); // Observe the draft count
```
Removed it.

Rajat-Dabade commented on app/components/drafts_buttton/index.tsx:
```diff
@@ -0,0 +1,31 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+/* eslint-disable max-nested-callbacks */
```
We don't this anymore, removed it.

Rajat-Dabade commented on app/screens/global_drafts/index.tsx:
```diff
@@ -0,0 +1,97 @@
+        return {flex: 1, marginTop};
+    }, [defaultHeight]);
+
+    const onBackPress = useCallback(() => {
```
Yup, that's a nice catch. On a tablet, if the back button is pressed, it should not pop the component from the navigation.

Rajat-Dabade commented:
@larkox I’ve addressed all the review comments. Thank you for highlighting the changes. They all make perfect sense. Could you please re-review it whenever you have some time? Much appreciated! 🙌

Rajat-Dabade commented:
> just 1 last change, in below flow
Scenarios 1:
Steps:
From Draft page, Edit a draft. Verify you are taken to the Channel/DM/GM
Edit the Draft in the Channel
Click on the < back Icon on the channel
This is taking user to Draft page. should it take user to Channels List page?

@yasserfaraazkhan, I have fixed this can you please retest it? Thank you.

yasserfaraazkhan commented:
/update-branch

yasserfaraazkhan reviewed (approved):
Verified on android and iOS device
- Drafts created in Channel/DM/GM 
- send/edit/delete actions on drafts
- swipe left on draft to delete
- drafts on public channel and then convert
- Remove server that has drafts and add it back to see no drafts are present
- Theme applied on the page.
- Drafts on multiple teams 
- Draft count updated when we leave channel, archive channel

Few observations,

- On android where we can see channels list page for fraction of a second, when we click on a particular Draft

https://github.com/user-attachments/assets/377a6d3e-4910-4931-b4ba-5344f1f18201

- Theme issues when Device is in Dark mode.

Steps:
1. select Onxy (dark theme)
2. create Draft in channel , 
3. change to Denim theme
4. go to channel and see the theme is applied
5. go to Draft, long press draft, Click on Edit

Actual: we see Dark theme section 

![Screenshot_20241211-151829](https://github.com/user-attachments/assets/8215a8a8-dde5-46d2-af53-d39f9c7d653c)
![Screenshot_20241211-151834](https://github.com/user-attachments/assets/fad34e6e-ff2e-4a21-b6c3-2c7fbc0a48dc)

enahum commented on app/queries/servers/drafts.ts:
```diff
@@ -30,3 +29,27 @@ export const queryDraft = (database: Database, channelId: string, rootId = '') =
     return v[0]?.observe() || of$(undefined);
 }
+
+export const queryAllDrafts = (database: Database, teamId: string) => {
```
Well is not all drafts is it?

Rajat-Dabade replied:
Yup, its correct, it for a specific team, will change the name.,

enahum commented on app/queries/servers/drafts.ts:
```diff
@@ -30,3 +29,27 @@ export const queryDraft = (database: Database, channelId: string, rootId = '') =
+    );
+};
+
+export const observeAllDrafts = (database: Database, teamId: string) => {
```
Same comment

enahum commented on app/screens/home/channel_list/categories_list/index.tsx:
```diff
@@ -72,6 +73,7 @@ const CategoriesList = ({hasChannels, iconPad, isCRTEnabled, moreThanOneTeam}: C
                         shouldHighlighActive={true}
                     />
                 }
+                <DraftsButtton/>
```
I wonder if this should only mount if there are any drafts

Rajat-Dabade replied:
The mounting is actually handled inside the component. Return null if there is not draft.

Rajat-Dabade replied:
Okay got it. Will change it.

enahum commented on app/screens/global_drafts/components/global_drafts_list/index.tsx:
```diff
@@ -0,0 +1,35 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+/* eslint-disable max-nested-callbacks */
```
Why this?

Rajat-Dabade replied:
Ah, not required. Will remove it.

enahum commented on app/screens/global_drafts/components/global_drafts_list/index.tsx:
```diff
@@ -0,0 +1,35 @@
+    };
+});
+
+export default React.memo(withDatabase(withTeamId(enhanced(GlobalDraftsList))));
```
Does it need memo? I'm not a fan with all these nested with unless actually makes sense

Rajat-Dabade replied:
Removed it.

enahum commented on app/screens/global_drafts/components/draft_empty_component/index.tsx:
```diff
@@ -0,0 +1,61 @@
+            <Image
+                source={draft_message_image}
+            />
+            <Text style={styles.title}>
+                {intl.formatMessage({id: 'drafts.empty.title', defaultMessage: 'No drafts at the moment'})}
+            </Text>
+            <Text style={styles.subtitle}>
+                {intl.formatMessage({id: 'drafts.empty.subtitle', defaultMessage: 'Any message you have started will show here.'})}
+            </Text>
```
Instead of text you can use formattedText component

enahum commented on app/screens/draft_options/edit_draft/index.tsx:
```diff
@@ -0,0 +1,78 @@
+                size={ICON_SIZE}
+                color={changeOpacity(theme.centerChannelColor, 0.56)}
+            />
+            <Text style={style.title}>{intl.formatMessage({id: 'draft.options.edit.title', defaultMessage: 'Edit draft'})}</Text>
```
Instead of Text you can use FormattedText

enahum commented on app/screens/draft_options/delete_draft/index.tsx:
```diff
@@ -0,0 +1,72 @@
+                size={ICON_SIZE}
+                color={changeOpacity(theme.centerChannelColor, 0.56)}
+            />
+            <Text style={style.title}>{intl.formatMessage({id: 'draft.options.delete.title', defaultMessage: 'Delete draft'})}</Text>
```
Use FormattedText instead of Text

enahum commented on app/screens/global_drafts/components/global_drafts_list/SwipeableDraft/index.tsx:
```diff
@@ -0,0 +1,129 @@
+                        color={theme.sidebarText}
+                        name='trash-can-outline'
+                        size={18}
+                        style={{}}
```
Remove?

Rajat-Dabade replied:
Yup, removed it.

enahum commented on app/screens/global_drafts/components/global_drafts_list/SwipeableDraft/index.tsx:
```diff
@@ -0,0 +1,129 @@
+                        size={18}
+                        style={{}}
+                    />
+                    <Text style={styles.deleteText}>{intl.formatMessage({
```
Use FormattedText

enahum commented on app/screens/global_drafts/components/global_drafts_list/global_drafts_list.tsx:
```diff
@@ -0,0 +1,146 @@
+
+    const tooltipContentStyle = useMemo(() => ({
+        borderRadius: 8,
+        maxWidth: isTablet ? 352 : 247,
```
Explain the magic numbers here

Rajat-Dabade replied:
Removed it as we no longer use it.

enahum commented on app/screens/global_drafts/components/global_drafts_list/global_drafts_list.tsx:
```diff
@@ -0,0 +1,146 @@
+    }, []);
+
+    const collapse = useCallback(() => {
+        if (Platform.OS === 'android') {
```
Why only android? And is seems this is called from the back button handler that is only available on Android, so seems unnecessary

Rajat-Dabade replied:
True, removing it.

enahum commented on app/screens/global_drafts/components/global_drafts_list/global_drafts_list.tsx:
```diff
@@ -0,0 +1,146 @@
+                    tooltipStyle={styles.tooltipStyle}
+                >
+                    <View
+                        style={{width: '100%'}}
```
Set this in the stylesheet

enahum commented on app/screens/global_drafts/components/global_drafts_list/global_drafts_list.tsx:
```diff
@@ -0,0 +1,146 @@
+
+    return (
+        <View
+            style={{flex: 1}}
```
Set this in the stylesheet

enahum commented on app/screens/draft_options/send_draft/send_draft.tsx:
```diff
@@ -0,0 +1,162 @@
+                size={ICON_SIZE}
+                color={changeOpacity(theme.centerChannelColor, 0.56)}
+            />
+            <Text style={style.title}>{intl.formatMessage({id: 'draft.options.send.title', defaultMessage: 'Edit draft'})}</Text>
```
Use FormattedText

enahum commented on app/screens/global_drafts/components/draft_empty_component/index.tsx:
```diff
@@ -0,0 +1,61 @@
+
+import React from 'react';
+import {useIntl} from 'react-intl';
+import {Image, Text, View} from 'react-native';
```
Somewhere else you are using Image from expo-image for the same purpose, which makes sense, change it here

Rajat-Dabade replied:
Done.

enahum commented on app/screens/global_drafts/components/global_drafts_list/tooltip.tsx:
```diff
@@ -0,0 +1,89 @@
+    onClose: () => void;
+}
+
+const logo = require('@assets/images/emojis/swipe.png');
```
Perhaps change the name of the variable so that it makes sense instead of logo

enahum commented on app/screens/global_drafts/index.tsx:
```diff
@@ -0,0 +1,99 @@
+
+import type {AvailableScreens} from '@typings/screens/navigation';
+
+const edges: Edge[] = ['left', 'right'];
```
Why not include bottom? Is it added elsewhere? I see top in the styles but not bottom. Or is it that we don't want it at all?

Rajat-Dabade replied:
This is what the design suggests https://www.figma.com/design/VbEx15Tpvi9mXhs9IEjJtq/MM-45558-Drafts-%26-Schedule-Messages?node-id=4013-70751&m=dev, also we have not used it in the Threads View.

enahum reviewed (changes requested):
Overall looks good, great job!

I have added a few comments

Rajat-Dabade commented on app/actions/local/draft.ts:
```diff
@@ -197,3 +211,83 @@ export async function updateDraftPriority(serverUrl: string, channelId: string,
+
+export async function parseMarkdownImages(markdown: string, imageMetadata: Dictionary<PostImage | undefined>) {
+    let match;
+    const imageRegex = /!\[.*?\]\((https:\/\/[^\s)]+)\)/g;
```
This is absolutely a nice catch, I will update the regex. Thank you 🫡 .

Rajat-Dabade commented on app/components/channel_info/avatar/index.tsx:
Changed the name to ProfileAavator, I think it makes more sense now.

Rajat-Dabade commented on app/components/channel_info/avatar/index.tsx:
```diff
@@ -0,0 +1,71 @@
+    const serverUrl = useServerUrl();
+
+    let uri = '';
+    if (author) {
```
True, we don't need this check here. Will remove it.

Rajat-Dabade commented on app/components/draft/draft.tsx:
```diff
@@ -0,0 +1,125 @@
+
+    return (
+        <TouchableHighlight
+            delayLongPress={200}
```
Not really required, have taken the reference from here: https://github.com/mattermost/mattermost-mobile/blob/b3d3ce3d8168c692ff406f35291ab89dbc713a06/app/components/post_list/post/post.tsx#L411

Will remove it.

Rajat-Dabade commented on app/components/draft/draft.tsx:
```diff
@@ -0,0 +1,125 @@
+            <View
+                style={style.container}
+            >
+                <ChannelInfo
```
This component will be used for scheduled posts component as well.

enahum replied:
Then assign a different name cause I think it clashes with the screen channel info

Rajat-Dabade replied:
Done.

Rajat-Dabade commented on app/components/draft/draft_post/draft_message/index.tsx:
```diff
@@ -0,0 +1,108 @@
+    const animatedStyle = useShowMoreAnimatedStyle(height, maxHeight, open);
+
+    const onLayout = useCallback((event: LayoutChangeEvent) => setHeight(event.nativeEvent.layout.height), []);
+    const onPress = () => setOpen(!open);
```
Updated it.

Rajat-Dabade commented on app/components/drafts_buttton/index.tsx:
```diff
@@ -0,0 +1,30 @@
+    };
+});
+
+export default React.memo(withDatabase(withTeamId(enhanced(DraftsButton))));
```
Sure will merged it. As I read more now about `withDatabase` itself is responsible for optimizing the database interactions and will typically ensure that the component is re-rendered only when relevant data changes so React.memo is not need. Will remove it.

enahum commented on app/actions/local/draft.ts:
```diff
@@ -197,3 +211,83 @@ export async function updateDraftPriority(serverUrl: string, channelId: string,
+    while ((match = imageRegex.exec(markdown)) !== null) {
+        const imageUrl = match[1];
+        // eslint-disable-next-line no-await-in-loop
+        imageMetadata[imageUrl] = await getImageMetadata(imageUrl);
```
Basically it can complete iterating while the valus on the object have not been set yet

Rajat-Dabade replied:
I have updated the code which ensures that the iteration waits for all asynchronous operation completes before proceeding. `Promise.all` guarantees that the `imageMetadata` object is updated after all `getImageMetadata` calls are resolved.

enahum commented on app/actions/local/draft.ts:
```diff
@@ -197,3 +211,93 @@ export async function updateDraftPriority(serverUrl: string, channelId: string,
+    // [^\s()<>]+ - Matches the main part of the URL, excluding spaces, parentheses, and angle brackets.
+    // (?:\([^\s()<>]+\))* - Allows balanced parentheses inside the URL path or query parameters.
+    // !\[.*?\]\((...)\) - Matches an image markdown syntax ![alt text](image url)
+    const imageRegex = /!\[.*?\]\((([a-zA-Z][a-zA-Z\d+\-.]*):\/\/[^\s()<>]+(?:\([^\s()<>]+\))*)\)/g;
```
hmmm, should this only allow http/https/ftp/ftps ? what about the extensions, do we want to do anything with that to only attempt those that are supported?

Rajat-Dabade replied:
0/5 on restricting to only http/https/ftp/ftps, since users can provide any valid URL, and as long as it returns an image, it should work. Also, I don't think extensions matter here because URLs don't always include extensions but can still return valid images.

enahum commented on app/actions/local/draft.ts:
```diff
@@ -197,3 +211,93 @@ export async function updateDraftPriority(serverUrl: string, channelId: string,
+    const imageRegex = /!\[.*?\]\((([a-zA-Z][a-zA-Z\d+\-.]*):\/\/[^\s()<>]+(?:\([^\s()<>]+\))*)\)/g;
+    const matches = Array.from(markdown.matchAll(imageRegex));
+
+    const promises = matches.map(async (match) => {
+        const imageUrl = match[1];
+        if (isValidUrl(imageUrl)) {
+            const metadata = await getImageMetadata(imageUrl);
+            imageMetadata[imageUrl] = metadata;
+        }
+    });
```
this does return anything so the promises will always be an array of undefined.

```suggestion
    const promises = matches.reduce<Array<Promise<MetadataReturnedByGetImageImagedataType>>((result, match) => {
        const imageUrl = match[1];
        if (isValidUrl(imageUrl)) {
            result.push(getImageMetadata(imageUrl));
        }
        return result;
    }, []);
```

getImageMetadata should include the url in the return object so that you can map it after or something like that

another option would be to use `for await` but this will make it sequential and not parallel

Rajat-Dabade replied:
Make sense.

enahum commented on app/screens/draft_options/index.tsx:
```diff
@@ -0,0 +1,113 @@
+    const theme = useTheme();
+    const styles = getStyleSheet(theme);
+    const snapPoints = useMemo(() => {
+        const bottomSheetAdjust = Platform.select({ios: 5, default: 20});
```
what is this adjust and why is it needed ?

Rajat-Dabade replied:
The adjustment is for height to add some padding at the bottom. Without this, it would touch the bottom of the screen.

enahum commented on app/screens/global_drafts/components/global_drafts_list/SwipeableDraft/index.tsx:
`SwipeableDraft` name convention does not match

Rajat-Dabade replied:
Changed it to `DraftSwipeActions`

enahum commented on app/utils/helpers.ts:
```diff
@@ -182,3 +182,12 @@ export function areBothStringArraysEqual(a: string[], b: string[]) {
     return areBothEqual;
 }
+
+export function isValidUrl(url: string): boolean {
```
there is an `isValidUrl` function in `app/utils/url` which may cause some confusion to devs, either be more specific, use or modify the one we already have. Also if you decide to keep this and change the name, make sure to at least move it the right file, and do add test coverage for it

Rajat-Dabade replied:
Changed the name to `isParseableUrl` as its reliance is on the URL constructor for validation.

enahum commented on app/utils/draft/index.ts:
add unit test for code coverage

enahum commented on app/utils/post/index.ts:
add unit test for code coverage

marianunez commented:
Adding @rahimrahman while Elias is out on PTO so we can complete the reviews here ASAP cc// @Rajat-Dabade

rahimrahman commented on app/actions/local/draft.ts:
```diff
@@ -197,3 +212,99 @@ export async function updateDraftPriority(serverUrl: string, channelId: string,
+                    resolve(null);
+                },
+                (error) => {
+                    logError('Failed to get image size', error);
```
Same here: Maybe "Failed getImageMetadata to get image size" ?

Rajat-Dabade replied:
Updated it.

rahimrahman commented on app/actions/local/draft.ts:
```diff
@@ -197,3 +212,99 @@ export async function updateDraftPriority(serverUrl: string, channelId: string,
+            );
+        });
+    } catch (error) {
+        width = 0;
+        height = 0;
```
Do we need this?

The width and height are set to 0 already.  What about the error?  Do we need try catch here?

Rajat-Dabade replied:
We don't need try catch. Removed it.

rahimrahman commented on app/actions/local/draft.ts:
```diff
@@ -197,3 +212,99 @@ export async function updateDraftPriority(serverUrl: string, channelId: string,
+        width = 0;
+        height = 0;
+    }
+    const match = url.match(/\.(\w+)(?=\?|$)/);
```
Can we add comment on what to expect from this regex?  Regex is tricky to understand sometime.

rahimrahman reviewed (commented):
I only got a few files down.  Will continue looking through.

But this is to start.

rahimrahman commented on app/components/draft_post_header/ProfileAvatar/index.tsx:
```diff
@@ -0,0 +1,68 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import {Image} from 'expo-image';
```
I feel like this is the same copy as https://github.com/mattermost/mattermost-mobile/blob/mobile-drafts/app/screens/gallery/footer/avatar/index.tsx.  Why not make it a common component?

Rajat-Dabade replied:
The styles are different, we need to pass the required styles for both the image and view container, should we do it for better have a separate component for it?

rahimrahman commented on app/components/draft_post_header/draft_post_header.tsx:
```diff
@@ -0,0 +1,162 @@
+
+    if (rootId) {
+        headerComponent = (
+            <View style={style.channelInfo}>
+                <FormattedText
+                    id='channel_info.thread_in'
+                    defaultMessage={'Thread in:'}
+                    style={style.category}
+                />
+                <View style={style.profileComponentContainer}>
+                    {profileComponent}
+                </View>
+            </View>
```
Can we create an inline component instead?  It seems between the three if-else conditions, the things that are different is the `id` and `defaultMessage.`

Rajat-Dabade replied:
Done.

rahimrahman commented on app/components/drafts_buttton/index.tsx:
```diff
@@ -0,0 +1,128 @@
+        switchToGlobalDrafts();
+    }), []);
+
+    const isActive = isTablet && shouldHighlighActive && !currentChannelId;
```
Spelling error.
```suggestion
    const isActive = isTablet && shouldHighlightActive && !currentChannelId;
```

Rajat-Dabade replied:
Updated it.

rahimrahman commented on app/components/drafts_buttton/index.tsx:
```diff
@@ -0,0 +1,128 @@
+    return (
+        <TouchableOpacity
+            onPress={handlePress}
+            testID='channel_list.drafts.button'
```
Curious, why is the testID has `channel_list` in it?

Rajat-Dabade replied:
This component is mounted where all the channels are listed. That's the reason why testID is `channel_list.drafts.button`. Also taken the reference from https://github.com/mattermost/mattermost-mobile/blob/4fedb846bca708031f89c200c9ad2b7d1b1cc3f0/app/components/threads_button/threads_button.tsx#L115 .

Rajat-Dabade replied:
Yup, not a common component.

rahimrahman commented on app/hooks/handle_send_message.ts:
```diff
@@ -0,0 +1,223 @@
+            };
+        }
+
+        createPost(serverUrl, post, postFiles);
```
Same here.  Should we worry if the post is created or not?

Rajat-Dabade replied:
This can also be taken up later.

rahimrahman commented on app/screens/home/channel_list/categories_list/categories_list.tsx:
```diff
@@ -0,0 +1,104 @@
+        return {maxWidth: withTiming(tabletWidth.value, {duration: 350})};
+    }, [isTablet, width]);
+
+    const content = useMemo(() => {
+        if (!hasChannels) {
+            return (<LoadChannelsError/>);
+        }
+
+        return (
+            <>
+                <SubHeader/>
+                {isCRTEnabled &&
+                    <ThreadsButton
+                        isOnHome={true}
+                        shouldHighlighActive={true}
+                    />
+                }
+                {draftsCount > 0 && (
+                    <DraftsButton
+                        draftsCount={draftsCount}
+                        currentChannelId={currentChannelId}
+                    />
+                )}
+                <Categories/>
+            </>
+        );
+    }, [currentChannelId, draftsCount, hasChannels, isCRTEnabled]);
```
Is there a huge benefit of memoizing this?  Wouldn't even draftsCount change would cause the entire content to be recreated? Would React default diff algorithm be efficient enough to handle the changes.  If draftsCount change, would it create a new memoized output or use the standard diffing on render?

rahimrahman replied:
React diffing algorithm knows what to change.  `content` gets re-assigned with JSX.element on any of the dependency changes.  So as an example, if draftsCount change, would ThreadsButton gets re-rendered as well?  Whereas React diffing would know not to do that.

Thus the reason of my query.  If React diffing works in useMemo as well, then I think we're all good here.  I would think so, but I'm not sure myself.

rahimrahman replied:
I was curious so I created a test myself:

https://codesandbox.io/p/sandbox/relaxed-shadow-5jskx4

The diffing works within useMemo.  As suspected.  But again, the React diffing algorithm should be able to handle this on it's own.  As changes to draftsCount would not cause ThreadsButton to re-render, similar to useMemo.  You can modify the same demo to mimic what you're doing.

Rajat-Dabade replied:
So I have created a test demo by modifying your changes.

https://codesandbox.io/p/sandbox/funny-diffie-hc88fj 

In the demo, I added a `console.log("Re-rendering")` inside `SomeOtherComponent` to verify if it re-renders. Upon clicking the "Press Me" button, the console logs `Re-rendering`, indicating that `SomeOtherComponent` is indeed re-rendering.

I also tested by removing the useMemo to see if React's diffing algorithm handles the re-rendering on its own. However, it doesn’t seem to prevent `SomeOtherComponent` from re-rendering in this case.

Let me know if you see anything I might be missing!

rahimrahman commented on app/screens/home/channel_list/categories_list/categories_list.tsx:
```diff
@@ -0,0 +1,104 @@
+        if (isTablet) {
+            tabletWidth.value = getTabletWidth(moreThanOneTeam);
+        }
+    }, [isTablet, moreThanOneTeam]);
```
```suggestion

    // tabletWidth is a sharedValue, so it's safe to ignore this warning
    // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [isTablet, moreThanOneTeam]);
```

rahimrahman commented on app/screens/global_drafts/components/global_drafts_list/DraftSwipeActions/index.tsx:
```diff
@@ -0,0 +1,130 @@
+import React, {useCallback, useEffect, useRef} from 'react';
+import {useIntl} from 'react-intl';
+import {Animated, DeviceEventEmitter, TouchableWithoutFeedback} from 'react-native';
+import {Swipeable} from 'react-native-gesture-handler';
```
There's a deprecation warning, should we use another component as suggested?

https://docs.swmansion.com/react-native-gesture-handler/docs/components/reanimated_swipeable

Rajat-Dabade replied:
Removed it and now using `ReanimatedSwipeable.

rahimrahman commented on app/screens/global_drafts/components/global_drafts_list/DraftSwipeActions/index.tsx:
```diff
@@ -0,0 +1,130 @@
+        });
+
+        return (
+            <TouchableWithoutFeedback
```
Should we start looking into using Pressable?

https://reactnative.dev/docs/pressable

Rajat-Dabade replied:
Sure.

rahimrahman commented on app/utils/draft/index.ts:
```diff
@@ -3,11 +3,13 @@
 import {t} from '@i18n';
 
 import type {IntlShape, MessageDescriptor} from 'react-intl';
+import type {Swipeable} from 'react-native-gesture-handler';
```
Deprecated notice.  Mentioned in another file.

Rajat-Dabade replied:
Removed it, and replaced it with `ReanimatedSwipeable`

rahimrahman commented on app/utils/url/index.test.ts:
```diff
@@ -27,3 +27,38 @@ describe('safeDecodeURIComponent', () => {
+    });
+
+    it('should return false for invalid URLs', () => {
+        expect(isParsableUrl('example')).toBe(false);
```
Is just 'example.com' a non-parsable URL?

Rajat-Dabade replied:
Yup, The `URL` constructor requires a protocol, without protocol it will throw an error and will be considered an invalid URL.

rahimrahman replied:
If you can, add it to your test (if you haven't) so somebody like me and others understand that protocol is required even though 'example.com' looks valid from the naked eye.

Rajat-Dabade replied:
Done. Added a test for it.

rahimrahman commented:
There's a few discrepancies when it comes to file naming convention.  This is something @larkox and I chatted in a PR once but I can't find that discussion.

Anyway, some file is index.tsx, some is send_draft.tsx.  If I remember correctly, the convention or at least the pattern is anything component/index.tsx would have the database observable, whereas the some_component_name.tsx is the component itself?

rahimrahman commented:
I've brought up with @marianunez about the need for E2E testing for any new feature.  This was brought up by @lindalumitchell && @lindy65 during a dev meeting.

This feature is a fine candidate for it, but understand the "urgency" in getting it through.  Nevertheless, I think there need to be negotiation with the QA team.

rahimrahman commented on app/screens/global_drafts/components/global_drafts_list/global_drafts_list.tsx:
```diff
@@ -0,0 +1,147 @@
+    },
+});
+
+const AnimatedFlatList = Animated.createAnimatedComponent(FlatList);
```
Should we consider using Animated.Flatlist since createAnimatedComponent(FlatList) is deprecated?

 @deprecated Please use `Animated.FlatList` component instead of calling `Animated.createAnimatedComponent(FlatList)` manually.

Rajat-Dabade replied:
Done.

rahimrahman commented on app/screens/global_drafts/components/global_drafts_list/global_drafts_list.tsx:
```diff
@@ -0,0 +1,147 @@
+        InteractionManager.runAfterInteractions(() => {
+            setTooltipVisible(true);
+        });
+    }, []);
```
Missing dependency.
```suggestion
    }, [tutorialWatched]);
```

Rajat-Dabade replied:
This should only run once on the mount, that the reason why the dependency was removed.

Rajat-Dabade replied:
Added comment and disabled eslint rule for dependency check.

rahimrahman replied:
Keep in mind that React 18 run useEffect twice in development + strict mode.  Although I don't think MM app is running on strict mode.  Which is a problem on it's own, but a different story.

rahimrahman commented on app/screens/global_drafts/components/global_drafts_list/DraftSwipeActions/index.tsx:
```diff
@@ -0,0 +1,130 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import React, {useCallback, useEffect, useRef} from 'react';
```
I notice the folder naming is different here than others.  DraftSwipeActions.  The pattern I'm seeing would have made this folder draft_swipe_actions.  I don't know if this is a rule/pattern but I think we ought to be consistent.

Rajat-Dabade replied:
Ah, will change it.

amyblais commented:
@Rajat-Dabade Kind note that v2.25 feature complete deadline is scheduled for Monday, Jan 13th.

Rajat-Dabade commented on app/hooks/handle_send_message.ts:
```diff
@@ -0,0 +1,223 @@
+    }, [sendingMessage, value, files, maxMessageLength]);
+
+    const handleReaction = useCallback((emoji: string, add: boolean) => {
+        handleReactionToLatestPost(serverUrl, emoji, add, rootId);
```
You're right. That's a valid point, and we can consider it. However, for this PR, I created a hook by copy-pasting the existing code to reuse it in multiple places. I think this can be discussed further and addressed in future refactoring.

Reference to the previous code: https://github.com/mattermost/mattermost-mobile/blob/8b5c50e6e8f721af06fd6c588ff4d54e5191f415/app/components/post_draft/send_handler/send_handler.tsx#L122

lindy65 commented:
> I've brought up with @marianunez about the need for E2E testing for any new feature. This was brought up by @lindalumitchell & @lindy65 during a dev meeting.
> 
> This feature is a fine candidate for it, but understand the "urgency" in getting it through. Nevertheless, I think there need to be negotiation with the QA team.

Although a new process is in discussion in [~testing-matters channel on hub](https://hub.mattermost.com/private-core/channels/testing-matters) details have not yet been finalised. I think tests for at least basic functionality should be in place before merging the feature.  I see [this set of tests](https://mattermost.atlassian.net/projects/MM?selectedItem=com.atlassian.plugins.atlassian-connect-plugin:com.kanoah.test-manager__main-project-page#!/v2/testCases) in Zephyr that were added by @yasserfaraazkhan. Once the feature is merged, these will either need to be run in Detox or added to and run in Rainforest

Rajat-Dabade commented on app/actions/local/draft.ts:
```diff
@@ -1,9 +1,24 @@
+    }
+};
 
 export async function updateDraftFile(serverUrl: string, channelId: string, rootId: string, file: FileInfo, prepareRecordsOnly = false) {
```
Will do this in separate refactoring PR.

rahimrahman replied:
@Rajat-Dabade Can you create a ticket for type refactoring on your changes?  While typing doesn't impact your implementation, typing is probably why the codebase is in typescript.

I'm a big fan of more detail typing vs just types.

```
/**
 * Base props for Draft
 */
  interface Draft {
      /** server url for the draft being written **/
      serverUrl: string;
      /** The channel of the draft **/
      channelId: string;
      /** I have no idea what rootId is for thus the documentation is great here! **/
      rootId: string;
  }
```

And if done right, hovering over a param will look like this:

![Screenshot 2025-01-09 at 08 22 00](https://github.com/user-attachments/assets/15c4d003-b859-4923-8d1c-576292027960)

Rajat-Dabade replied:
Created a ticket for tracking: https://mattermost.atlassian.net/browse/MM-62495

Rajat-Dabade commented:
@rahimrahman Addressed review comments also taken care of the file names as you suggested. Can you please re-review it? Let me know if there is anything else. Thank you 🙌

Rajat-Dabade commented on app/hooks/handle_send_message.ts:
```diff
@@ -0,0 +1,223 @@
+        }
+
+        return messageLength > 0;
+    }, [sendingMessage, value, files, maxMessageLength]);
```
Done.

Rajat-Dabade commented on app/screens/global_drafts/components/global_drafts_list/tooltip.tsx:
```diff
@@ -0,0 +1,85 @@
+    },
+});
+
+const DraftTooltip = ({onClose}: Props) => {
```
Changed the name of the file to `draft_tooltip.tsx`

rahimrahman commented:
> Although a new process is in discussion in [~testing-matters channel on hub](https://hub.mattermost.com/private-core/channels/testing-matters) details have not yet been finalised. I think tests for at least basic functionality should be in place before merging the feature. I see [this set of tests](https://mattermost.atlassian.net/projects/MM?selectedItem=com.atlassian.plugins.atlassian-connect-plugin:com.kanoah.test-manager__main-project-page#!/v2/testCases) in Zephyr that were added by @yasserfaraazkhan. Once the feature is merged, these will either need to be run in Detox or added to and run in Rainforest

@marianunez @Rajat-Dabade maybe create a ticket for detox test and get it through within the next few cycles?

rahimrahman commented on app/utils/draft/index.ts:
```diff
@@ -9,7 +9,7 @@ import {CODE_REGEX} from '@constants/autocomplete';
 
 import type {IntlShape, MessageDescriptor} from 'react-intl';
-import type {Swipeable} from 'react-native-gesture-handler';
+import type {SwipeableMethods} from 'react-native-gesture-handler/lib/typescript/components/ReanimatedSwipeable';
```
should it be 

```
from 'react-native-gesture-handler/ReanimatedSwipeable';
```

?

Rajat-Dabade replied:
That nice catch, I changed it.

Rajat-Dabade commented:
> @marianunez @Rajat-Dabade maybe create a ticket for detox test and get it through within the next few cycles?

We already have ticket created for this: https://mattermost.atlassian.net/browse/MM-60500

rahimrahman reviewed (approved):
Great job with all the work on this PR.  And thank you for working through all the changes.

This is a big PR, and I'd love to see it done in smaller chunks, with proper testing attached to new files.

yasserfaraazkhan commented:
Thank you @Rajat-Dabade 

- As expected we can't send in private channel, when User is removed
- As expected we can't send in archived channel
- As expected we can send in GM/DM when they are closed.
- We see error when we send the Draft in a Channel that was changed to Read-only
<img width="525" alt="image" src="https://github.com/user-attachments/assets/3c84f2ab-9d44-46ff-aef9-031bf62f5e38" />
- Able to create, delete and send Drafts, in different channel, teams multiple servers.

Rajat-Dabade commented:
@yasserfaraazkhan Thank you for testing. 

Regarding the issue you mentioned: I have created a ticket for it: https://mattermost.atlassian.net/browse/MM-62513. It's not a blocker. Will take care of this issue in subsequent PR.

asaadmahmood reviewed (approved):
LGTM, we just need to update the drafts icon to:
`icon-send-outline`.

https://mattermost.github.io/compass-icons/

![CleanShot 2025-01-13 at 16 23 27@2x](https://github.com/user-attachments/assets/b11826e6-8e77-42c6-9a25-b4e569c90562)

Rajat-Dabade commented:
Done @asaadmahmood Updated the Icon.

<img width="409" alt="Screenshot 2025-01-13 at 5 02 38 PM" src="https://github.com/user-attachments/assets/48cfcfd0-a39b-43b7-a1b4-a6821359efae" />

mattermost-build commented:
Cherry pick is scheduled.
